### PR TITLE
[Mailer] Add support for multiple mailers

### LINKED ISSUE #32535: [Mailer] Can create many mailers like in SwiftMailer
**Description**  
Hello all,
the new Mailer component released in 4.3 is to replace the SwiftMailer component. Everything works as it should. I have a huge project that connects to many databases, many REST servers and many SMTP servers. Traditionally, according to the documentation, I use SwiftMailer to send e-mails and it looks like this:

```
swiftmailer:
    default_mailer: mailerone
    mailers:
        mailerone:
            transport: '%mailer_one_transport_protocol%'
            host:      '%mailer_one_transport_host%'
            username:  '%mailer_one_transport_user%'
            password:  '%mailer_one_transport_password%'
            spool: { type: memory }
            logging:   '%kernel.debug%'
        mailersecond:
            transport: '%mailer_second_transport_protocol%'
            host:      '%mailer_second_transport_host%'
            username:  '%mailer_second_transport_user%'
            password:  '%mailer_second_transport_password%'
            spool: { type: memory }
            logging:   '%kernel.debug%'
        mailerthird:
            transport: '%mailer_third_transport_protocol%'
            host:      '%mailer_third_transport_host%'
            username:  '%mailer_third_transport_user%'
            password:  '%mailer_third_transport_password%'
            spool: { type: memory }
            logging:   '%kernel.debug%'
        mailerfourth:
            transport: '%mailer_fourth_transport_protocol%'
            host:      '%mailer_fourth_transport_host%'
            username:  '%mailer_fourth_transport_user%'
            password:  '%mailer_fourth_transport_password%'
            spool: { type: memory }
            logging:   '%kernel.debug%'

```

Unfortunately, I can not transform the implementation into a new component because it lacks the declaration of many Transports...

```
framework:
    mailer:
        dsn: '%env(MAILER_DSN)%'
```

**Example**  
I do not know if this is a good idea, but I think many people also have or will have a similar problem.

```
framework:
    mailer:
        default_mailer: mailerone
        mailers:
            mailerone: '%env(MAILER_ONE_DSN)%'
            mailersecond: '%env(MAILER_SECOND_DSN)%'
```

Sorry for my english, is so bad. Thank you for reading my problem and at the same time suggesting adding functionality (actually, in my opinion huge).

### PR DESCRIPTION
**Pull request by fabpot**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | closes #32535
| License       | MIT
| Doc PR        | -

This adds the possibility to define several email transports.

If you only have one email transport,  nothing changes:

```yaml
framework:
    mailer:
        dsn: '%env(MAILER_DSN)%'
```

But if you need more than one, use the `transports` entry instead:

```yaml
framework:
    mailer:
        transports:
            main: '%env(MAILER_DSN)%'
            important: '%env(MAILER_DSN_IMPORTANT)%'
```

Then, when sending an email via the `Mailer`, pass the mailer name explicitly (by default, the first one is used):

```php
// use the first "main" transport
$mailer->send($email);

// or use the "important" one
$email->getHeaders()->addTextHeader('X-Transport', 'important');
$mailer->send($email);
```

The web profiler now displays the name and the shorten DSN.

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 1]
```diff
@@ -316,6 +316,10 @@ public function load(array $configs, ContainerBuilder $container)
             $this->registerHttpClientConfiguration($config['http_client'], $container, $loader, $config['profiler']);
         }
 
+        if ($this->mailerConfigEnabled = $this->isConfigEnabled($container, $config['mailer'])) {
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: This is the same as #33408

File: src/Symfony/Component/Mailer/Bridge/Sendgrid/Tests/Transport/SendgridTransportFactoryTest.php [hunk 2]
```diff
@@ -76,4 +76,9 @@ public function unsupportedSchemeProvider(): iterable
             'The "foo" scheme is not supported for mailer "sendgrid". Supported schemes are: "api", "smtp", "smtps".',
         ];
     }
+
+    public function incompleteDsnProvider(): iterable
+    {
+        yield [new Dsn('api', 'sendgrid')];
+    }
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Sendgrid/Tests/Transport/SendgridTransportFactoryTest.php]: Not related, but avoids a skipped test.

**Comment by OskarStark** [comment]
You wrote mailgun instead of important in the PR header example code

**Comment by fabpot** [comment]
@OskarStark fixed :)

File: src/Symfony/Component/Mailer/Messenger/SendEmailMessage.php [hunk 3]
```diff
@@ -19,18 +19,24 @@
...
     private $envelope;
 
     /**
      * @internal
      */
-    public function __construct(RawMessage $message, SmtpEnvelope $envelope = null)
+    public function __construct(RawMessage $message, SmtpEnvelope $envelope = null, string $transport = null)
     {
         $this->message = $message;
         $this->envelope = $envelope;
```
Comment (michaljusiega) [inline src/Symfony/Component/Mailer/Messenger/SendEmailMessage.php]: `$transport` variable assignment missing?

File: src/Symfony/Component/Mailer/Mailer.php [hunk 4]
```diff
@@ -24,21 +23,32 @@
  */
 class Mailer implements MailerInterface
 {
-    private $transport;
+    private $transports;
     private $bus;
     private $dispatcher;
 
-    public function __construct(TransportInterface $transport, MessageBusInterface $bus = null, EventDispatcherInterface $dispatcher = null)
+    public function __construct(Transports $transports, MessageBusInterface $bus = null, EventDispatcherInterface $dispatcher = null)
```
Comment (Koc) [inline src/Symfony/Component/Mailer/Mailer.php]: Usually this kind of services named as `TransportLocator`

Comment (fabpot) [reply src/Symfony/Component/Mailer/Mailer.php]: As you know, I'm not a big fan of using a pattern name in a class name. `Transports` sounds much better to me.

**Comment by ro0NL** [comment]
> by default, the first one is used)

is this convention sensible? should we support an explicit option? i.e.

```yaml
default_transport: some
````

**Comment by OskarStark** [comment]
> is this convention sensible? should we support an explicit option? i.e.

Yes much better, doctrine or messenger are doing the same 👍

File: src/Symfony/Component/Mailer/Transports.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ */
+class Transports
+{
+    private $transports;
+    private $default;
+
+    /**
+     * @param TransportInterface[] $transports
+     */
+    public function __construct(iterable $transports)
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Transports.php]: Couldn't we use sth like: `__construct(Transport ...$transports)`?

**Comment by fabpot** [comment]
In which case this convention would be a problem? The sort does not matter except for the first one which is the default.

File: src/Symfony/Component/Mailer/Transports.php (hunk 5 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Mailer/Transports.php]: Look, the keys of the iterator are used. And the check is done on the return type of getTransport. That's way enough for the use case.

**Comment by fabpot** [comment]
In which case this convention would be a problem? The sort does not matter except for the first one which is the default. Less configuration is better, right?

**Comment by ro0NL** [comment]
> The sort does not matter except for the first one which is the default. 

so the sorting matters :)

> Less configuration is better, right?

yes :) but im not sure we should ignore the configuration part in  "convention over configuration" otherwise it's just a magic convention.

**Comment by fabpot** [comment]
I fail to see the magic here. The first one is the default one. That's all.

**Comment by ro0NL** [comment]
both work :) it's a doc issue eventually. To me the magic is, updating the ordering (i.e. to "alpha order") might change the default implicitly.

however, that's the case anyway as long as someones relies on the "first is default" convention, without knowing about `default_transport`.

Let's keep it simple then :+1:

**Comment by ro0NL** [comment]
note the messenger configuration does enforce 100% explicit configuration:

https://github.com/symfony/symfony/blob/bcb91eadf04631cb13834312c1123bc9b3c4276c/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L1129-L1130

**Comment by fabpot** [comment]
New (simpler/better?) implementation without BC breaks.

**Review by nicolas-grekas — APPROVED** [review]

**Pull request merged**